cmd/bootstrap/async_replication.go

```diff
@@ -0,0 +1,268 @@
+package main
+
+import (
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/sjmudd/stopwatch"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	apiv1alpha1 "github.com/percona/percona-server-mysql-operator/api/v1alpha1"
+	"github.com/percona/percona-server-mysql-operator/pkg/mysql"
+	"github.com/percona/percona-server-mysql-operator/pkg/replicator"
+)
+
+func bootstrapAsyncReplication() error {
+	timer := stopwatch.NewNamedStopwatch()
+	err := timer.AddMany([]string{"clone", "total"})
+	if err != nil {
+		return errors.Wrap(err, "add timers")
+	}
+	timer.Start("total")
+
+	defer func() {
+		timer.Stop("total")
+		log.Printf("bootstrap finished in %f seconds", timer.ElapsedSeconds("total"))
+	}()
+
+	svc := os.Getenv("SERVICE_NAME_UNREADY")
+	mysqlSvc := os.Getenv("SERVICE_NAME")
+	peers, err := lookup(svc)
+	if err != nil {
+		return errors.Wrap(err, "lookup")
+	}
+	log.Printf("Peers: %v", peers.List())
+
+	exists, err := lockExists()
+	if err != nil {
+		return errors.Wrap(err, "lock file check")
+	}
+	if exists {
+		log.Printf("Waiting for bootstrap.lock to be deleted")
+		if err = waitLockRemoval(); err != nil {
+			return errors.Wrap(err, "wait lock removal")
+		}
+	}
+	primary, replicas, err := getTopology(peers)
+	if err != nil {
+		return errors.Wrap(err, "select donor")
+	}
+	log.Printf("Primary: %s Replicas: %v", primary, replicas)
+
+	fqdn, err := getFQDN(mysqlSvc)
+	if err != nil {
+		return errors.Wrap(err, "get FQDN")
+	}
+	log.Printf("FQDN: %s", fqdn)
+
+	podHostname, err := os.Hostname()
+	if err != nil {
+		return errors.Wrap(err, "get hostname")
+	}
+
+	podIp, err := getPodIP(podHostname)
+	if err != nil {
+		return errors.Wrap(err, "get pod IP")
+	}
+	log.Printf("PodIP: %s", podIp)
+
+	primaryIp, err := getPodIP(primary)
+	if err != nil {
+		return errors.Wrap(err, "get primary IP")
+	}
+	log.Printf("PrimaryIP: %s", primaryIp)
+
+	donor, err := selectDonor(fqdn, primary, replicas)
+	if err != nil {
+		return errors.Wrap(err, "select donor")
+	}
+	log.Printf("Donor: %s", donor)
+
+	if donor == "" || donor == fqdn || primary == fqdn || primaryIp == podIp {
+		return nil
+	}
+
+	log.Printf("Opening connection to %s", podIp)
+	operatorPass, err := getSecret(apiv1alpha1.UserOperator)
+	if err != nil {
+		return errors.Wrapf(err, "get %s password", apiv1alpha1.UserOperator)
+	}
+
+	db, err := replicator.NewReplicator("operator", operatorPass, podIp, mysql.DefaultAdminPort)
```

Comment:
can we have some wait logic or try to reconnect to DB because now we can have
`bootstrap failed: connect to db: ping database: dial tcp 10.0.1.7:33062: connect: connection refused`
and as the result unneeded restart

Reply:
do you see this with GR or async?

Reply:
GR